python/lsst/afw/typehandling/_GenericMap.py

```diff
@@ -73,3 +113,49 @@ def pop(self, key, default=None):
                 return default
             else:
                 raise
+
+
+MutableGenericMap.register(str, MutableGenericMapS)
+MutableMapping.register(MutableGenericMapS)
+
+
+class AutoKeyMeta(TemplateMeta):
+    """A metaclass for abstract mappings whose key type is implied by their
+    constructor arguments.
+
+    This metaclass requires that the mapping have a `dict`-like constructor,
+    i.e., it takes a mapping or an iterable of key-value pairs as its first
+    positional parameter.
+
+    This class differs from `~lsst.utils.TemplateMeta` only in that the dtype
+    (or equivalent) constructor keyword is optional. If it is omitted, the
+    class will attempt to infer it from the first argument.
+    """
+
+    def __call__(cls, *args, **kwargs):  # noqa N805, non-self first param
+        if len(cls.TEMPLATE_PARAMS) != 1:
+            raise ValueError("AutoKeyMeta requires exactly one template parameter")
+        dtypeKey = cls.TEMPLATE_PARAMS[0]
+        dtype = kwargs.get(dtypeKey, None)
+
+        # Try to infer dtype if not provided
+        # Assume homogeneous keys; constructor will raise TypeError if they're not
+        if dtype is None and len(args) >= 1:
+            inputData = args[0]
+            if inputData:
+                if isinstance(inputData, Mapping):
+                    # mapping to copy
+                    firstKey = iter(inputData.keys()).__next__()
+                    dtype = type(firstKey)
+                elif not isinstance(inputData, str):
```

Comment:
`if`, `elif`, but where's the `else`? What happens if `inputData` is a string?

Reply:
The map constructor will fail, then, same as if you tried to make a `dict` from a string.

Reply:
Maybe it will be clear in refactoring, but a comment about the missing `else` may be worthwhile.